### PR TITLE
Implement ByteBuffer-based APIs for BitShuffle

### PR DESCRIPTION
**Pull request by maropu**
This pr added the implementations of ByteBuffer-based APIs for `BitShuffle` (#155).

### CODE REVIEW

**Comment by xerial** [comment]
I think we also should support `byte[]` backed ByteBuffer. Throwing exception for non-direct byte buffer can be confusing.

**Comment by xerial** [comment]
Ah. The current Snappy methods already rejecting non-direct buffers and I haven't seen any complaint with this. So this is ok. Never mind.

File: src/main/java/org/xerial/snappy/BitShuffle.java [hunk 1]
```diff
@@ -43,6 +44,46 @@
...
+            throw new SnappyError(SnappyErrorCode.NOT_A_DIRECT_BUFFER, "destination is not a direct buffer");
+        }
+
+        // input: input[pos(), limit())
+        // output: shuffled
+        int uPos = input.position();
+        int uLen = input.remaining();
+        int typeSize = type.getTypeSize();
+        if (uLen % typeSize != 0) {
+            throw new IllegalArgumentException("input length must be a multiple of a given type size");
```
Comment (xerial) [inline src/main/java/org/xerial/snappy/BitShuffle.java]: given data type size

File: src/main/java/org/xerial/snappy/BitShuffle.java [hunk 2]
```diff
@@ -113,6 +154,46 @@
...
+            throw new SnappyError(SnappyErrorCode.NOT_A_DIRECT_BUFFER, "destination is not a direct buffer");
+        }
+
+        // input: input[pos(), limit())
+        // output: shuffled
+        int uPos = shuffled.position();
+        int uLen = shuffled.remaining();
+        int typeSize = type.getTypeSize();
+        if (uLen % typeSize != 0) {
+            throw new IllegalArgumentException("length of input shuffled data must be a multiple of a given type size");
```
Comment (xerial) [inline src/main/java/org/xerial/snappy/BitShuffle.java]: ditto

File: src/test/java/org/xerial/snappy/BitShuffleTest.java [hunk 3]
```diff
@@ -26,11 +26,250 @@
...
+        ByteBuffer directBuf = ByteBuffer.allocateDirect(64);
+
+        // Tests for BitShuffle.bitShuffle()
+        try {
+            BitShuffle.bitShuffle(heapBuf, BitShuffleType.BYTE, directBuf);
+            fail("no expected exception happened");
+        }
+        catch (SnappyError e) {
+            Assert.assertTrue(e.errorCode == SnappyErrorCode.NOT_A_DIRECT_BUFFER);
+            System.err.println("XXX:" + e.getMessage());
```
Comment (xerial) [inline src/test/java/org/xerial/snappy/BitShuffleTest.java]: Unnecessary line.

File: src/main/java/org/xerial/snappy/BitShuffle.java (hunk 1 above)
Comment (xerial) [inline src/main/java/org/xerial/snappy/BitShuffle.java]: or BitShuffleType

**Comment by xerial** [comment]
And run `make linux64` to build linux binaries to pass the test on Travis CI.

**Comment by maropu** [comment]
okay, fixed.

**Comment by xerial** [comment]
TravisCI is now green :) Thanks for the improvement.

**Comment by xerial** [comment]
@maropu Thanks for renaming. This looks much simpler!

**Pull request merged**